src/Shared/FileUtilities.cs

```diff
@@ -393,6 +392,68 @@ private static bool IsUNCPath(string path)
             return isUNC || path.IndexOf(@"\\?\globalroot", StringComparison.OrdinalIgnoreCase) != -1;
         }
 
+#if !CLR2COMPATIBILITY
+        /// <summary>
+        /// Gets the exact case used on the file system for an existing file or directory.
+        /// </summary>
+        /// <remarks>
+        /// This method should only be called on Windows.
+        /// </remarks>
+        /// <param name="path">A relative or absolute path.</param>
+        /// <returns>The full path using the correct case if the path exists. Otherwise, the provied path.</returns>
+        internal static string GetExactPath(string path)
+        {
+            // DirectoryInfo accepts either a file path or a directory path, and most of its properties work for either.
+            // However, its Exists property only works for a directory path, so we need an explicit file existence check.
+            var directory = new DirectoryInfo(path);
+            if (!File.Exists(path) && !directory.Exists)
+            {
+                // If the path doesn't exist, just return the full path.
+                return path;
+            }
+
+            // We know exactly how long the string will be, so avoid unnecessary intermediate allocations.
+            var chars = new char[directory.FullName.Length];
+            var currentLocation = chars.Length;
+
+            DirectoryInfo parentDirectory = directory.Parent;
+            while (parentDirectory != null)
+            {
+                FileSystemInfo entry = parentDirectory.EnumerateFileSystemInfos(directory.Name).First();
+
+                // Copy the correctly-cased path part to our working character array.
+                string pathPart = entry.Name;
+                currentLocation -= pathPart.Length;
+                pathPart.CopyTo(0, chars, currentLocation, pathPart.Length);
```

Comment:
Does this backward-array-copying thing work in languages with more-complex capitalization rules?

I think this will fail, though maybe Windows will fail beforehand?

* On a German machine
* Name a project `thing_with_SS.proj`
* `msbuild.exe thing_with_ß.proj`